### ISSUE TITLE
Acquiring an exclusive Navigator LockManager failed

### ISSUE THREAD

**Issue by supersu-man** [supabase/supabase-js#936]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

Angular throws an error saying 'Acquiring an exclusive Navigator LockManager lock immediately failed'.
This is a new angular project created with the angular version 17.
The error is not shown in my other project which is created a month ago.

## To Reproduce

1. Create a new angular project and install '@supabase/supabase-js'
2. Create environment files and add key and url of supabase.
3. add `supabaseClient = createClient(environment.url, environment.key)` to a component (app component)

## Expected behavior

This error should not be thrown and should be able to use all the methods without errors.

## Screenshots

![Error](https://github.com/supabase/supabase-js/assets/35676632/edd0fd65-dac6-4df2-8752-7987a367a4ee)

## System information

- OS: Windows
- Browser: Brave v1.61.109
- Version of supabase-js: 2.39.2
- Version of Node.js: 20.10.0

## Additional context

Could be a duplicate of [this issue](https://github.com/supabase/gotrue-js/issues/830) from gotrue-js.
Though I was able to signin using provider 'google' I wasn't able to retrieve session after I signin (throws an error saying Invalid API key) though my api keys and url are correct. I am assuming the error is with storage.

**Comment by Klunk75** [comment]
Hi there,
I noticed the same problem.
After some test I found that the issue should be related to the merge of this Pull Request in @supabase/gotrue-js version 2.62.0:
https://github.com/supabase/gotrue-js/pull/807
If you try to downgrade @supabase/gotrue-js to version 2.61.0 (the previous one), the problem should disappear.
Please Supabase Team can you check and try to solve this issue?
Thank you!

Paolo

**Comment by supersu-man** [comment]
> If you try to downgrade @supabase/gotrue-js to version 2.61.0 (the previous one), the problem should disappear. 

Thank you @Klunk75 for the workaround.

For anyone who wants to fix it temporarily, add this below devDependencies in package.json

```
"overrides": {
    "@supabase/supabase-js": {
      "@supabase/gotrue-js": "2.61.0"
    }
  }
```
and run `npm i`

**Comment by hf** [comment]
Is this error just being logged, or is it actually creating a problem for you?

**Comment by supersu-man** [comment]
> Is this error just being logged, or is it actually creating a problem for you?

Sorry for the late reply. I don't think I am facing any issue, it's just the errors being logged.

**Comment by dimeloper** [comment]
In case you use `pnpm` as your dependency manager and you're facing the same issue, you can fix it by adding the following in your package.json:

```
"pnpm": {
    "overrides": {
      "@supabase/supabase-js>@supabase/gotrue-js": "2.61.0"
    }
  },
```

**Comment by mikelhamer** [comment]
seeing this with supabase 2.39.7, angular 17.2.1

**Comment by linhub15** [comment]
Seeing this with supabase 2.39.7
Reproduced with Chrome 121.0.6167.184 (Official Build) (64-bit)
Reproduced with Safari 16.6 (18615.3.12.11.2)
Unable to reproduce with Firefox 123.0 (64-bit).

```ts
// login.tsx
await supabase.auth.signInWithOAuth({
  provider: "google",
});

// index.tsx
const session = await supabase.auth.getSession();  // session is null
```
![image](https://github.com/supabase/supabase-js/assets/10420994/424ab769-8b08-47d2-a166-c101f25bf2e4)

Overriding gotrue-js to `"2.61.0"` fixed it.

**Comment by danielehrhardt** [comment]
> > If you try to downgrade @supabase/gotrue-js to version 2.61.0 (the previous one), the problem should disappear.
> 
> Thank you @Klunk75 for the workaround.
> 
> For anyone who wants to fix it temporarily, add this below devDependencies in package.json
> 
> ```
> "overrides": {
>     "@supabase/supabase-js": {
>       "@supabase/gotrue-js": "2.61.0"
>     }
>   }
> ```
> 
> and run `npm i`

maybe you can merge that

**Comment by matheo** [comment]
@vv001 the override worked for me with `@supabase/supabase-js@2.40.0` but for version `2.41.1` it seems that `gotrue` was renamed to `authjs` and then I updated the override and it worked!

for `pnpm`:
```
"overrides": {
    "@supabase/supabase-js>@supabase/auth-js": "2.61.0"
}
```
and for `npm` should look like:
```
"overrides": {
  "@supabase/supabase-js": {
    "@supabase/auth-js": "2.61.0"
  }
}
```

**Comment by vv001** [comment]
Hi martheo, that did the trick. Thank you very much!

**Comment by ibarraandre98** [comment]
@matheo your solution worked for me, thanks!

**Comment by aroman25** [comment]
@matheo - thanks! this worked for me and then I updated supabase for the signInAnonymously and now getting this error again, can you work your magic again for the latest @supabase/supabase-js 2.42.5?

**Comment by wszgrcy** [comment]
> @matheo - thanks! this worked for me and then I updated supabase for the signInAnonymously and now getting this error again, can you work your magic again for the latest @supabase/supabase-js 2.42.5?

same bug.....want use `signInAnonymously`.

**Comment by matheo** [comment]
@aroman25 @wszgrcy weird because I just upgraded it and the same override worked for me :thinking: 
in your lock file is there a `@supabase/auth-js` different than `2.61.0`?

**Comment by ThirzaNL** [comment]
I had the same issue as @aroman25 and @wszgrcy , and I've solved it using a [patch](https://gist.github.com/ThirzaNL/4f263f85fc934a926b8196468e909d4b).

Let me preface by saying that I don't know the inner workings of Supabase at all, and I have not tested it thoroughly, so use it at your own risk.

Basically, it just removes the navigatorLock parts, as that's seemingly causing the error.

**Comment by aroman25** [comment]
@wszgrcy - Yeah, that worked, also not a fan of it being experimental, but seems like v18 will have zoneless coming out next month

@matheo - 2.61.0 doesn't have the ability to signInAnonymously( )

@ThirzaNL - thanks, I'll wait for a fix, I can get by for now

**Comment by alcaidio** [comment]
Same issue with Angular 17.3.8 and pnpm, i dont know why but i cant fix the bug...even with the overrides in package.json

**Comment by nicetomytyuk** [comment]
Any update on it?

I have     "@supabase/supabase-js": "^2.43.4" and auth-js is 2.64.2 and still having that issue:

![image](https://github.com/supabase/supabase-js/assets/30237552/f6776f48-2e0f-49ad-bb70-c454088ab627)

**Comment by macgors** [comment]
I have the same problem with supabase-js 2.43.5 and angular 18.0.5

**Comment by francocores** [comment]
@macgors maybe this could fix your problem

for me it's working on angular 18 with ssr but I need do a workaround on localhost (I'm not sure if this problem happens on https environment):

my service
`
@Injectable({
  providedIn: 'root',
})
export class SupabaseService {
  private supabase?: SupabaseClient
  _session: AuthSession | null = null

  constructor() {
    if (typeof window === 'undefined' || !window.location || createClient === undefined) return
    this.supabase = createClient(environment.supabaseUrl, environment.supabaseKey)
  }

  categories() {
    return this.supabase?.from('angular_test_categories').select(`id, name`)
  }
}
`

my package.json
"dependencies": {
    ...
    "@angular/core": "^18.0.0",
    "@angular/ssr": "^18.0.7",
    ...
    "@supabase/supabase-js": "2.39.2",
  },
  "overrides": {
    "@supabase/supabase-js": {
      "@supabase/gotrue-js": "2.61.0"
    }
  },

**Comment by francocores** [comment]
@macgors maybe this could fix your problem

for me it's working on angular 18 with ssr but I need do a workaround on localhost (I'm not sure if this problem happens on https environment):

my service
```typescript
@Injectable({
  providedIn: 'root',
})
export class SupabaseService {
  private supabase?: SupabaseClient
  _session: AuthSession | null = null


  constructor() {
    this.createClient()
  }

  createClient() {
    if (typeof window === 'undefined' || createClient === undefined) return
    this.supabase = createClient(environment.supabaseUrl, environment.supabaseKey)
  }

  categories() {
    return this.supabase?.from('angular_test_categories').select(`id, name`)
  }
}
```

my package.json
```node
"dependencies": {
    ...
    "@angular/core": "^18.0.0",
    "@angular/ssr": "^18.0.7",
    ...
    "@supabase/supabase-js": "2.39.2",
  },
  "overrides": {
    "@supabase/supabase-js": {
      "@supabase/gotrue-js": "2.61.0"
    }
  },
```

**Comment by mungo-pod** [comment]
Bumping, still can't get it to work even with overrides, Was on Angular 18 and downgraded to 17.3.8 per one of the comments above, but no luck. here's the loadout. 

```json 
"dependencies": {
    "@supabase/supabase-js": "2.39.2",
  },
  "devDependencies": {
    "supabase": "^1.167.1",
  },
  "overrides": {
    "@supabase/supabase-js": {
      "@supabase/gotrue-js": "2.61.0"
    }
  }

**Comment by rileythomp** [comment]
@mungo-pod See [this comment](https://github.com/supabase/supabase-js/issues/936#issuecomment-2027940513) above, the override needs to be on `"@supabase/auth-js"` now.

**Comment by RyannGalea** [comment]
I still have not been able to resolve this using Yarn. latest versions of both

**Comment by Picoquio** [comment]
It's been eight months....

**Comment by Picoquio** [comment]
It's been eight months. Do you guys really expect everyone that sets up a new project on angular to stumble upon the helpful replies of this particular thread?

**Comment by wszgrcy** [comment]
Downgrading is a very bad approach
Because the new version has new features and bug fixes
We can temporarily downgrade to solve this problem
But it's been several months now, and I think a formal repair should be carried out

**Comment by TRUSTMEIMJEDI** [comment]
Any plans on fixing this issue in near future?

**Comment by machariaz** [comment]
I am using Angular 18 and the issue still persist even after  trying to fix using the above recommendations

**Comment by AlexisMelo** [comment]
> @vv001 the override worked for me with `@supabase/supabase-js@2.40.0` but for version `2.41.1` it seems that `gotrue` was renamed to `authjs` and then I updated the override and it worked!
> 
> for `pnpm`:
> 
> ```
> "overrides": {
>     "@supabase/supabase-js>@supabase/auth-js": "2.61.0"
> }
> ```
> 
> and for `npm` should look like:
> 
> ```
> "overrides": {
>   "@supabase/supabase-js": {
>     "@supabase/auth-js": "2.61.0"
>   }
> }
> ```

same problem as everyone with
- angular 18.1.3
- supabase-js 2.45.1

this comment worked

**Comment by PrettyJuice** [comment]
With Supabase 2.45.2 and Angular 18.2.1, the problem still persist.

> FYI this is what resolved it for me with yarn.
> 
> ```
>   "resolutions": {
>     "@supabase/auth-js": "2.61.0"
>   }
> ```

This works, I hope it is a short time temporarily solution

**Comment by kleytonroan** [comment]
> With Supabase 2.45.2 and Angular 18.2.1, the problem still persist.
> 
> > FYI this is what resolved it for me with yarn.
> > ```
> >   "resolutions": {
> >     "@supabase/auth-js": "2.61.0"
> >   }
> > ```
> 
> This works, I hope it is a short time temporarily solution

I tried all the alternatives here, but none of them worked for me, since I'm using the new feature called anonymous authentication, which prevents me from changing the resolutions for version 2.61.0.

I had no time to test others versions

**Comment by alcaidio** [comment]
Why doesn’t the Supabase team address these issues? The problem has been ongoing for quite some time. I love Supabase, but I’m also a big fan of Angular. It's disappointing that the Angular community doesn’t seem to be a priority for Supabase.

**Comment by Bforsyth1234** [comment]
I would also like this issue to get some love.

**Comment by Picoquio** [comment]
> Why doesn’t the Supabase team address these issues? The problem has been ongoing for quite some time. I love Supabase, but I’m also a big fan of Angular. It's disappointing that the Angular community doesn’t seem to be a priority for Supabase.

It's sad to see most of the new players treating Angular as a second class citizen when it's megaparsecs superior to React in almost every aspect

**Comment by francocores** [comment]
> @macgors maybe this could fix your problem
> 
> for me it's working on angular 18 with ssr but I need do a workaround on localhost (I'm not sure if this problem happens on https environment):
> 
> my service
> 
> ```ts
> @Injectable({
>   providedIn: 'root',
> })
> export class SupabaseService {
>   private supabase?: SupabaseClient
>   _session: AuthSession | null = null
> 
> 
>   constructor() {
>     this.createClient()
>   }
> 
>   createClient() {
>     if (typeof window === 'undefined' || createClient === undefined) return
>     this.supabase = createClient(environment.supabaseUrl, environment.supabaseKey)
>   }
> 
>   categories() {
>     return this.supabase?.from('angular_test_categories').select(`id, name`)
>   }
> }
> ```
> 
> my package.json
> 
> ```js
> "dependencies": {
>     ...
>     "@angular/core": "^18.0.0",
>     "@angular/ssr": "^18.0.7",
>     ...
>     "@supabase/supabase-js": "2.39.2",
>   },
>   "overrides": {
>     "@supabase/supabase-js": {
>       "@supabase/gotrue-js": "2.61.0"
>     }
>   },
> ```

I'm using the version 2.44.4 on Angular 18 and it's working without override

package.json
````
"dependencies": {
    "@angular/animations": "^18.0.0",
    "@angular/cdk": "^18.0.6",
    "@angular/common": "^18.0.0",
    "@angular/compiler": "^18.0.0",
    "@angular/core": "^18.0.0",
    "@angular/forms": "^18.0.0",
    "@angular/platform-browser": "^18.0.0",
    "@angular/platform-browser-dynamic": "^18.0.0",
    "@angular/platform-server": "^18.0.0",
    "@angular/router": "^18.0.0",
    "@angular/ssr": "^18.0.7",
    "@supabase/supabase-js": "2.44.4",
    "chart.js": "4.4.4",
    "date-fns": "^3.6.0",
    "express": "^4.18.2",
    "primeflex": "^3.3.1",
    "primeicons": "^7.0.0",
    "primeng": "^17.18.5",
    "rxjs": "~7.8.0",
    "swiper": "11.1.11",
    "tslib": "^2.3.0",
    "zone.js": "~0.14.3"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "^18.0.7",
    "@angular/cli": "^18.0.7",
    "@angular/compiler-cli": "^18.0.0",
    "@types/chart.js": "^2.9.41",
    "@types/express": "^4.17.17",
    "@types/jasmine": "~5.1.0",
    "@types/node": "^18.18.0",
    "autoprefixer": "^10.4.19",
    "jasmine-core": "~5.1.0",
    "karma": "~6.4.0",
    "karma-chrome-launcher": "~3.2.0",
    "karma-coverage": "~2.2.0",
    "karma-jasmine": "~5.1.0",
    "karma-jasmine-html-reporter": "~2.1.0",
    "postcss": "^8.4.39",
    "typescript": "~5.4.2"
  }
````

**Comment by on3dd** [comment]
Cannot find a workaround for this issue to work with Angular SSR. Shame on the devs.

**Comment by JulienLecoq** [comment]
> megaparsecs superior to React in almost every aspect

It's sad indeed, but Angular is not megaparsecs superior to React in almost every aspect. The verbosity of Angular is a big drawback imho. You have to double imports things such as components, enum and so on when you need to use them inside your components.. You have to add to every component ``standalone: true`` when you want to use them as "regular" component like in any modern framework (although that should not be true after v19). You have to add a selector name to every component. There is too many ways of doing things, ``styles, styleUrls, styleUrl`` just for styling a component whereas a single and standard way could be sufficient. Overall, I think the whole ``@Component()`` approach could be optional or removed and we could just replace it by a single render() method inside of classes... Furthermore, currently the framework is in a big transition which makes it hard for new programmer to learn it. The docs are not always up to date, they are some recommendations that should not be true anymore, they even made an RFC for that recently. And, for a very long time, the framework was RXJS oriented which is a big nonsense imho. It just adds a lot of complexity for no reasons in most cases. Fortunately, there is a shifting toward signals.

So no, Angular is not megaparsecs superior to React in almost every aspect.

**Comment by on3dd** [comment]
I am currently using the following solution to overcome this issue using Angular 18 with SSR, maybe it will help someone.

supabase.service.ts
```ts

@Injectable({ providedIn: 'root' })
export class SupabaseService {
  private _initFinished$ = new ReplaySubject<void>(1);
  readonly initFinished$ = this._initFinished$.asObservable();

  public client!: SupabaseClient;

  initClient() {
    this.client = createClient(<...>);
    this._initFinished$.next();
  }
}
```

auth.service.ts
```ts
@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly supabaseService = inject(SupabaseService);

  constructor() {
    this.supabaseService.initFinished$.pipe(first()).subscribe(() => {
      this.supabaseService.client.auth.getSession().then(({ data }) => {
        <...>
      });
    });
  }
```

app.component.ts
```ts
@Component({ ... })
export class AppComponent {
  private readonly supabaseService = inject(SupabaseService);

  constructor() {
    afterNextRender(() => {
      this.supabaseService.initClient();
    });
  }
}
```

**Comment by lennybakkalian** [comment]
> > megaparsecs superior to React in almost every aspect
> 
> It's sad indeed, but Angular is not megaparsecs superior to React in almost every aspect. The verbosity of Angular is a big drawback imho. You have to double imports things such as components, enum and so on when you need to use them inside your components.. You have to add to every component `standalone: true` when you want to use them as "regular" component like in any modern framework (although that should not be true after v19). You have to add a selector name to every component. There is too many ways of doing things, `styles, styleUrls, styleUrl` just for styling a component whereas a single and standard way could be sufficient. Overall, I think the whole `@Component()` approach could be optional or removed and we could just replace it by a single render() method inside of classes... Furthermore, currently the framework is in a big transition which makes it hard for new programmer to learn it. The docs are not always up to date, they are some recommendations that should not be true anymore, they even made an RFC for that recently. And, for a very long time, the framework was RXJS oriented which is a big nonsense imho. It just adds a lot of complexity for no reasons in most cases. Fortunately, there is a shifting toward signals.
> 
> So no, Angular is not megaparsecs superior to React in almost every aspect.

There's someone yapping around.

Angular has put a strong focus on being backwards compatible. That's why deprecated methods/fields still exist.
standalone is not (yet) set to true by default for this reason.

The same applies to styleUrls and styleUrl. In react you have to add a library for every little thing. in the end you have a fat dependency on different maintainers.

a render() method? angular is very different from react because template and logic are separated. look at the react components nowadays. long tailwind classes + state logic + props nightmare all in one file. terrible to maintain.

**Comment by delongZhai** [comment]
This is a nightmare bug to work with.

I am using Angular because I was familiar with it, even for SSR, I still use Angular. But the problem with all above now, since it has been siting for a long time, is that some of the key feature just can't be used.

Anonymous Sign-In is a important functionality. But this temp fix disable the ability to use it.
`
"overrides": {
    "@supabase/supabase-js": {
      "@supabase/auth-js": "2.61.0"
    }
  }
`

I hope that the temp fix doesn't have to be rely upon for too long because it will hurt the longer you wait

**Comment by on3dd** [comment]
@delongZhai Hey, have you managed to get this to work with SSR? I posted my solution above that solved the main problem for me, but I quickly ran into a limitation of that solution when I tried to make an AuthGuard based on `user$` from `AuthService`. That would have worked fine if it weren't for the problem that brought us all here. 

I'm almost feel desperate. It looks like I'll have to disable SSR after all, because the devs are too lazy to fix this problem even in 10 months. What a shame. I don't think I'll ever try Supabase on my projects again after this.

**Comment by delongZhai** [comment]
hey @on3dd 

I would recommend you to use `isPlatformBrowser` fn to check if it's being run in the browser again.

Here is my entire code 

`
@Component({
  selector: 'app-root',
  standalone: true,
  imports: [
    RouterOutlet,
    MobileMenuComponent,
    HeaderComponent,
    FooterComponent,
  ],
  templateUrl: './app.component.html',
  styleUrl: './app.component.scss',
})
export class AppComponent {
  title = 'cbd-store';

  constructor(
    @Inject(PLATFORM_ID) private platformId: Object,
    private supabase: SupabaseService
  ) {}

  ngOnInit(): void {
    if (isPlatformBrowser(this.platformId) && !this.supabase.client) {
      this.supabase.initialize();
    }
  }
}
`

**Comment by delongZhai** [comment]
> @delongZhai Hey, have you managed to get this to work with SSR? I posted my solution above that solved the main problem for me, but I quickly ran into a limitation of that solution when I tried to make an AuthGuard based on `user$` from `AuthService`. That would have worked fine if it weren't for the problem that brought us all here.
> 
> I'm almost feel desperate. It looks like I'll have to disable SSR after all, because the devs are too lazy to fix this problem even in 10 months. What a shame. I don't think I'll ever try Supabase on my projects again after this.

There's still a resolution I can think of, depending on how complex your system is, which is to ditch the supabase browser client, but to use supabase server client entirely, which is a bit more nasty. But hey nextjs is doing that, and I think if you are fine with it, you can as well.

**Comment by on3dd** [comment]
> hey @on3dd
> 
> I would recommend you to use `isPlatformBrowser` fn to check if it's being run in the browser again.
> 
> Here is my entire code
> 
> ` @component({ selector: 'app-root', standalone: true, imports: [ RouterOutlet, MobileMenuComponent, HeaderComponent, FooterComponent, ], templateUrl: './app.component.html', styleUrl: './app.component.scss', }) export class AppComponent { title = 'cbd-store';
> 
> constructor( @Inject(PLATFORM_ID) private platformId: Object, private supabase: SupabaseService ) {}
> 
> ngOnInit(): void { if (isPlatformBrowser(this.platformId) && !this.supabase.client) { this.supabase.initialize(); } } } `

@delongZhai Thank you, PLATFORM_ID is something I had already forgotten about. 

> > @delongZhai Hey, have you managed to get this to work with SSR? I posted my solution above that solved the main problem for me, but I quickly ran into a limitation of that solution when I tried to make an AuthGuard based on `user$` from `AuthService`. That would have worked fine if it weren't for the problem that brought us all here.
> > I'm almost feel desperate. It looks like I'll have to disable SSR after all, because the devs are too lazy to fix this problem even in 10 months. What a shame. I don't think I'll ever try Supabase on my projects again after this.
> 
> There's still a resolution I can think of, depending on how complex your system is, which is to ditch the supabase browser client, but to use supabase server client entirely, which is a bit more nasty. But hey nextjs is doing that, and I think if you are fine with it, you can as well.
Sounds like something worth considering.

I don't have as much experience with Angular SSR as I do with Next.js, but doing all this things on the server sounds like a real pain, especially in Angular because you won't have those API handlers of Next.js that can fix things a bit 😖

**Comment by delongZhai** [comment]
> > hey @on3dd
> > I would recommend you to use `isPlatformBrowser` fn to check if it's being run in the browser again.
> > Here is my entire code
> > ` @component({ selector: 'app-root', standalone: true, imports: [ RouterOutlet, MobileMenuComponent, HeaderComponent, FooterComponent, ], templateUrl: './app.component.html', styleUrl: './app.component.scss', }) export class AppComponent { title = 'cbd-store';
> > constructor( @Inject(PLATFORM_ID) private platformId: Object, private supabase: SupabaseService ) {}
> > ngOnInit(): void { if (isPlatformBrowser(this.platformId) && !this.supabase.client) { this.supabase.initialize(); } } } `
> 
> @delongZhai Thank you, PLATFORM_ID is something I had already forgotten about.
> 
> > > @delongZhai Hey, have you managed to get this to work with SSR? I posted my solution above that solved the main problem for me, but I quickly ran into a limitation of that solution when I tried to make an AuthGuard based on `user$` from `AuthService`. That would have worked fine if it weren't for the problem that brought us all here.
> > > I'm almost feel desperate. It looks like I'll have to disable SSR after all, because the devs are too lazy to fix this problem even in 10 months. What a shame. I don't think I'll ever try Supabase on my projects again after this.
> > 
> > 
> > There's still a resolution I can think of, depending on how complex your system is, which is to ditch the supabase browser client, but to use supabase server client entirely, which is a bit more nasty. But hey nextjs is doing that, and I think if you are fine with it, you can as well.
> > Sounds like something worth considering.
> 
> I don't have as much experience with Angular SSR as I do with Next.js, but doing all this things on the server sounds like a real pain, especially in Angular because you won't have those API handlers of Next.js that can fix things a bit 😖

That's correct, for server-side code, I've been especially worried about. And I will continue to make sure my code can be migrated to client side the moment it's resolved. But apparently that's the only workaround for developer who not only need anonymous login and also need to use supabase, and Angular

**Comment by delongZhai** [comment]
For anyone who still blocked by this issue, please prioritize it by submitting a ticket to supabase support in your projects dashboard. The longer we wait, the longer we are blocked by it

**Comment by JulienLecoq** [comment]
> > It's sad indeed, but Angular is not megaparsecs superior to React in almost every aspect. The verbosity of Angular is a big drawback imho. You have to double imports things such as components, enum and so on when you need to use them inside your components.. You have to add to every component `standalone: true` when you want to use them as "regular" component like in any modern framework (although that should not be true after v19). You have to add a selector name to every component. There is too many ways of doing things, `styles, styleUrls, styleUrl` just for styling a component whereas a single and standard way could be sufficient. Overall, I think the whole `@Component()` approach could be optional or removed and we could just replace it by a single render() method inside of classes... Furthermore, currently the framework is in a big transition which makes it hard for new programmer to learn it. The docs are not always up to date, they are some recommendations that should not be true anymore, they even made an RFC for that recently. And, for a very long time, the framework was RXJS oriented which is a big nonsense imho. It just adds a lot of complexity for no reasons in most cases. Fortunately, there is a shifting toward signals.
> > So no, Angular is not megaparsecs superior to React in almost every aspect.
> 
> There's someone yapping around.
> 
> Angular has put a strong focus on being backwards compatible. That's why deprecated methods/fields still exist. standalone is not (yet) set to true by default for this reason.
> 
> The same applies to styleUrls and styleUrl. In react you have to add a library for every little thing. in the end you have a fat dependency on different maintainers.
> 
> a render() method? angular is very different from react because e.g. template and logic are separated for a good reason. look at the react components nowadays. long tailwind classes + state logic + props nightmare all in one file. terrible to maintain.
> 
> On a more neutral note, it is difficult to compare these two things. one is a framework and the other is a library (unless you are talking about e.g. nextjs)
> 
> But I also have to admit that as far as ssr is concerned, nextjs is clearly superior.

Yep, I understand why a lot of those "bad things" are still there, even though Angular has a strong focus on being backwards compatible that still does not make it megaparsecs superior to React in almost every aspect, come on...

I don't really see how ``styleUrls`` and ``styleUrl`` is related to react having the need to add libraries for more things than Angular.


> > It's sad indeed, but Angular is not megaparsecs superior to React in almost every aspect. The verbosity of Angular is a big drawback imho. You have to double imports things such as components, enum and so on when you need to use them inside your components.. You have to add to every component `standalone: true` when you want to use them as "regular" component like in any modern framework (although that should not be true after v19). You have to add a selector name to every component. There is too many ways of doing things, `styles, styleUrls, styleUrl` just for styling a component whereas a single and standard way could be sufficient. Overall, I think the whole `@Component()` approach could be optional or removed and we could just replace it by a single render() method inside of classes... Furthermore, currently the framework is in a big transition which makes it hard for new programmer to learn it. The docs are not always up to date, they are some recommendations that should not be true anymore, they even made an RFC for that recently. And, for a very long time, the framework was RXJS oriented which is a big nonsense imho. It just adds a lot of complexity for no reasons in most cases. Fortunately, there is a shifting toward signals.
> > So no, Angular is not megaparsecs superior to React in almost every aspect.
> 
> There's someone yapping around.
> 
> Angular has put a strong focus on being backwards compatible. That's why deprecated methods/fields still exist. standalone is not (yet) set to true by default for this reason.
> 
> The same applies to styleUrls and styleUrl. In react you have to add a library for every little thing. in the end you have a fat dependency on different maintainers.
> 
> a render() method? angular is very different from react because e.g. template and logic are separated for a good reason. look at the react components nowadays. long tailwind classes + state logic + props nightmare all in one file. terrible to maintain.
> 
> On a more neutral note, it is difficult to compare these two things. one is a framework and the other is a library (unless you are talking about e.g. nextjs)
> 
> But I also have to admit that as far as ssr is concerned, nextjs is clearly superior.

Yep, I understand why a lot of those "bad things" are still there, even though Angular has a strong focus on being backwards compatible that still does not make it megaparsecs superior to React in almost every aspect, come on...

I highly disagree on the take "long tailwind classes + state logic + props nightmare all in one file is terrible to maintain". That's why component are there. You can just decompose when it starts to be too big. And it's not mandatory to put everything in one file or use tailwind, you can use css modules and put your css in another file if you prefer to do so (at least with NextJS).
I still often myself not put everything into one file in Angular but this is mainly because it is long/verbose to create components in Angular and also because Angular is creating a supplementary tag every time you create a component, which makes styling afterwards more difficult and impact perf compared to other modern frameworks :(

I agree that on some points they're not comparable since they don't have the same amount of features. It would be more fair to compare to a full fledge React framework such as NextJS. But the component authoring format can be compared and the React one is less verbose, at least for the creation of a basic component (without JSX and so on involved).

I don't say Angular is shit and shouldn't be used, in fact Angular becomes much better release after release, less verbose and more enjoyable to use. But saying that it is megaparsecs superior to React in almost every aspect is such a lie.

**Comment by kangmingtay** [comment]
Hi everyone, apologies for the lack of replies by the team in this thread - the auth team isn't the most familiar with angular here but we want to understand the root cause of this issue so we can fix it properly without having to do something hacky like downgrading the version or asking angular users to remove the lock. 

> Angular throws an error saying 'Acquiring an exclusive Navigator LockManager lock immediately failed'.

Specifically, is Angular not able to access the Navigator Lock API? Based off a quick search, it seems like Angular 17 / 18 is compatible with the Navigator Lock API.

@hf has also looked into this earlier this year and we suspect that it's an issue with an unhandled promise rejection - https://github.com/supabase/auth-js/issues/830#issuecomment-1893666402 

we would really appreciate any help from the community with understanding more about the inner workings of angular here!

**Comment by delongZhai** [comment]
> hey @kangmingtay , thank for the reply. I don't have much knowledge about navigator lock api. Neither can I find much detail within Angular community.

But a quick look into this, it forces me to think if there's anything that's wrong with using navigator lock api with zone.js. Zone.js is a basically how Angular magically update the view whenever there the app runs change detection.

https://stackoverflow.com/a/52392036 

I wonder if using zoneless would make any difference. If it does, perhaps for angular team, we need to move it off the main thread to use web worker? I'm writing to just propose potential problems. I am not certain that it is the case since zone.js has been a while for a long time. But can you also explain how that navigator lock work? Currently, zonejs is running as default, with known issue for SSR, and very few have zoneless implemented. If you have time to look into it, that would be great. Thank you

https://angular.dev/guide/experimental/zoneless

a topic with zone.js pollution. https://blog.bitsrc.io/angular-mitigating-zone-pollution-in-angular-22734b19576e

**Comment by hf** [comment]
The way Zone.js works with promises is that it patches the native Promise API so it tracks the "execution context" in very hacky ways.

The [Navigator Locks](https://developer.mozilla.org/en-US/docs/Web/API/LockManager/request) API uses *native* promises from the looks of it. Meaning -- the browser's _internal_ Promise is being used, not the patched up Zone.js one. This is why the context is being lost and the weird log messages appear.

As far as I am aware, there's really no way to fix this. One way to get rid of the issue is to just not use the navigator lock -- which I strongly do not recommend as it would cause other subtle bugs in your application where people are randomly logged out. To do that, set the `lock` Auth option to a noop function in `createClient`.

**Comment by hf** [comment]
Actually maybe one thing that could help is if we were to wrap the lock promise with the patched promise. I know that Dexie's zone.js equivalent likes that.

**Comment by bobbyg603** [comment]
Hey @hf thanks for taking a look at this. I just updated to the latest supabase-js with the updated auth-js 2.67.3. Unfortunately this error still occurs. Any other ideas?

**Comment by davidecampello** [comment]
What I can't understand is if today it's possible to use supabase with angular / ionic project

**Comment by Vatsov** [comment]
I'm using Supabase 2.47.10, and the problem persists.

**Comment by Klunk75** [comment]
> What I can't understand is if today it's possible to use supabase with angular / ionic project or if this problem can produce bugs

Ciao Davide,
I'm using Supabase with an Ionic/Angular project and now it's working with no errors and no issues.
I'm using the latest version of all the frameworks ("@ionic/angular": "8.4.1" - "@angular/cli": "19.0.5" - "@supabase/supabase-js": "2.47.10") and I don't see this issue anymore.

Paolo

**Comment by nik2208** [comment]
hey @bobbyg603 @cyborgx99 @klunk75 all of u can confirm that that configuration does not rise the error? 'couse in my case it still persists
```ts
"dependencies": {
    "@angular/animations": "^19.0.5",
    "@angular/common": "^19.0.5",
    "@angular/compiler": "^19.0.5",
    "@angular/core": "^19.0.5",
    "@angular/forms": "^19.0.5",
    "@angular/platform-browser": "^19.0.5",
    "@angular/platform-browser-dynamic": "^19.0.5",
    "@angular/router": "^19.0.5",
    "@ng-bootstrap/ng-bootstrap": "^18.0.0",
    "@popperjs/core": "^2.11.8",
    "@supabase/supabase-js": "^2.47.10",
    "angularx-qrcode": "^19.0.0",
    "bootstrap": "^5.3.3",
    "bootstrap-icons": "^1.11.3",
    "jwt-decode": "^4.0.0",
    "ng2-pdf-viewer": "^10.4.0",
    "ngx-scanner-qrcode": "^1.7.3",
    "pdf-lib": "^1.17.1",
    "rxjs": "~7.8.0",
    "tslib": "^2.3.0",
    "zone.js": "~0.15.0"
  },
  
  export const appConfig: ApplicationConfig = {
  providers: [provideExperimentalZonelessChangeDetection(), /*provideZoneChangeDetection({ eventCoalescing: true }),*/ provideRouter(routes), provideHttpClient(withFetch())]
};
```

```bash
NavigatorLockAcquireTimeoutError: Acquiring an exclusive Navigator LockManager lock "lock:sb-xnthnwftctgsbyituxyp-auth-token" immediately failed
```

**Comment by rabiedadi** [comment]
The fix didn't help 
- `Angular` version: 19.1.0
- `Supabase` version: 2.48.0
![Image](https://github.com/user-attachments/assets/4f0d7b52-9162-4559-8858-696256f21e8b)

**Comment by badi-jabbour** [comment]
I'm having the same issue. Any fix for this?

```
"dependencies": {
    "@angular/animations": "^19.1.0",
    "@angular/common": "^19.1.0",
    "@angular/compiler": "^19.1.0",
    "@angular/core": "^19.1.0",
    "@angular/forms": "^19.1.0",
    "@angular/platform-browser": "^19.1.0",
    "@angular/platform-browser-dynamic": "^19.1.0",
    "@angular/router": "^19.1.0",
    "@supabase/supabase-js": "^2.48.1",
    "echarts": "^5.6.0",
    "ngx-echarts": "^19.0.0",
    "rxjs": "~7.8.0",
    "tslib": "^2.3.0",
    "zone.js": "~0.15.0"
  },
```

**Comment by chrismclarke** [comment]
> The fix didn't help
> 
> * `Angular` version: 19.1.0
> * `Supabase` version: 2.48.0
>   ![Image](https://github.com/user-attachments/assets/4f0d7b52-9162-4559-8858-696256f21e8b)

I'm using the same version and workaround, I still see the error logged but the SDK still works to send/receive data from my database so I'm treating it more as a warning than an error for now.

I also tested going fully [zoneless](https://angular.dev/guide/experimental/zoneless) on my application which did indeed remove the error entirely (although quite a big refactor for a large codebase, so planning to do so slowly over time with the other workaround in place)

**Comment by aliarshad9691** [comment]
> [@vv001](https://github.com/vv001) the override worked for me with `@supabase/supabase-js@2.40.0` but for version `2.41.1` it seems that `gotrue` was renamed to `authjs` and then I updated the override and it worked!
> 
> for `pnpm`:
> 
> ```
> "overrides": {
>     "@supabase/supabase-js>@supabase/auth-js": "2.61.0"
> }
> ```
> 
> and for `npm` should look like:
> 
> ```
> "overrides": {
>   "@supabase/supabase-js": {
>     "@supabase/auth-js": "2.61.0"
>   }
> }
> ```

This solved the issue for me.

**Comment by itsxHrisi** [comment]
Im trying to fix this issue by adding the override but it doesnt have any effect, any other solution? This is my code, am i doing it wrong?

![Image](https://github.com/user-attachments/assets/07317866-d11e-441a-8750-ec126784eb5b)

**Comment by Klunk75** [comment]
Hi all,
I solved removing Zone.js from my Angular project, moving to Zoneless change detection with the use of Signals.
Here what I did:

1. In main.ts add this: "provideExperimentalZonelessChangeDetection()" in the providers array
2. In polyfills.ts comment these lines: "import './zone-flags';" and "import 'zone.js';"
3. Convert any variable in the project that updates the user interface to a Signal

I hope this can help!
Kind regards,

Paolo

**Comment by nik2208** [comment]
> Convert any variable in the project that updates the user interface to a Signal
@Klunk75 when u say 'any' what do u exactly mean?
u're not using anymore interpolation or binding or anything? only signals?

**Comment by krabochki** [comment]
guys, thank you very much, we managed to solve the problem. 
to sum up:

in package.json:
```
"overrides": {
  "@supabase/supabase-js": {
    "@supabase/auth-js": "2.61.0"
  }
}
```
in app.config.ts in appConfig var .providers 
+ `provideExperimentalZonelessChangeDetection()`
- `provideZoneChangeDetection({ eventCoalescing: true })`

in angular.json in polyfills mentions removed  `"zone.js"`

at the end `npm i` ofc

**Comment by bobbyg603** [comment]
> guys, thank you very much, we managed to solve the problem. to sum up:
> 
> in package.json:
> 
> ```
> "overrides": {
>   "@supabase/supabase-js": {
>     "@supabase/auth-js": "2.61.0"
>   }
> }
> ```
> 
> in app.config.ts in appConfig var .providers add `provideExperimentalZonelessChangeDetection()` remove `provideZoneChangeDetection({ eventCoalescing: true })`
> 
> in angular.json in polyfills mentions removed `"zone.js"`
> 
> at the end `npm i` ofc

I think you can remove that override in package.json, zoneless change detection should be sufficient

**Comment by esthersoftwaredev** [comment]
I have tried the suggested solution of doing the following things: 

1. in `app.config.ts` replacing `provideZoneChangeDetection({ eventCoalescing: true })` with `provideZoneChangeDetection({ eventCoalescing: true })` 
2. in `angular.json.polyfills` removing `zone.js`
3. in `package.json` adding:
```
"overrides": {
  "@supabase/supabase-js": {
    "@supabase/auth-js": "2.61.0"
  }
}
```
4. stopping the dev server running `npm i` starting again, even at some point I believe I deleted the `node_modules` & `package.lock.json` and the `npm i` again.

And yes it is true - this did remove the error (attaching the prod version the dev version was worded a bit differently):

![Image](https://github.com/user-attachments/assets/95470fd4-0d75-41e5-a681-1b8524a6b318)

But when I actually proceed to sign in using OAuth (GitHub in my case) I got a whole bunch of other errors:

![Image](https://github.com/user-attachments/assets/cd66fb10-b4a0-4c92-b159-09164e3f0151)

Sooo... I discarded all those changes, never committed them to my repo... got back to my original code - and NOW there are no errors at all, none of them - I log in and out GitHub without any errors... 

- my `package.json` does not include `overrides` at all
- my `angular.json` still has `zone.js` in `polyfills` 
- and my `app.config.ts` has `provideZoneChangeDetection({ eventCoalescing: true }),`

my complete `package.json`

```
{
  "name": "ng-vision",
  "version": "0.1.0-beta",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "build:prod": "ng build --configuration production",
    "watch": "ng build --watch --configuration development",
    "test": "ng test"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "^19.1.0",
    "@angular/common": "^19.1.0",
    "@angular/compiler": "^19.1.0",
    "@angular/core": "^19.1.0",
    "@angular/forms": "^19.1.0",
    "@angular/platform-browser": "^19.1.0",
    "@angular/platform-browser-dynamic": "^19.1.0",
    "@angular/router": "^19.1.0",
    "@primeng/themes": "^19.0.6",
    "@supabase/supabase-js": "^2.48.1",
    "primeicons": "^7.0.0",
    "primeng": "^19.0.6",
    "rxjs": "~7.8.0",
    "tslib": "^2.3.0",
    "zone.js": "~0.15.0"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "^19.1.0",
    "@angular/cli": "^19.1.0",
    "@angular/compiler-cli": "^19.1.0",
    "@types/jasmine": "~5.1.0",
    "jasmine-core": "~5.5.0",
    "karma": "~6.4.0",
    "karma-chrome-launcher": "~3.2.0",
    "karma-coverage": "~2.2.0",
    "karma-jasmine": "~5.1.0",
    "karma-jasmine-html-reporter": "~2.1.0",
    "typescript": "~5.7.2"
  }
}

```
Environment
- windows
- vs code
- chrome

that was a puzzling experience...

**Comment by Eliot6001** [comment]
I've got the same problem in my project, i'm using VITE+React.js. 

`"dependencies": {
    "@hookform/resolvers": "^3.9.0",
    "@radix-ui/react-alert-dialog": "^1.1.6",
    "@radix-ui/react-avatar": "^1.1.0",
    "@radix-ui/react-dialog": "^1.1.1",
    "@radix-ui/react-dropdown-menu": "^2.1.1",
    "@radix-ui/react-label": "^2.1.0",
    "@radix-ui/react-separator": "^1.1.0",
    "@radix-ui/react-slot": "^1.1.2",
    "@radix-ui/react-toast": "^1.2.1",
    "@radix-ui/react-toggle": "^1.1.0",
    "@radix-ui/react-toggle-group": "^1.1.0",
    "@supabase/supabase-js": "^2.45.1",
    "@tailwindcss/typography": "github:tailwindcss/typography",
    "@tiptap/extension-image": "^2.7.1",
    "@tiptap/extension-link": "^2.7.1",
    "@tiptap/extension-underline": "^2.7.1",
    "@tiptap/pm": "^2.7.1",
    "@tiptap/react": "^2.7.1",
    "@tiptap/starter-kit": "^2.7.1",
    "@uploadthing/react": "^7.3.0",
    "antd": "^5.20.2",
    "axios": "^1.7.7",
    "class-variance-authority": "^0.7.0",
    "clsx": "^2.1.1",
    "draft-js": "^0.11.7",
    "global": "^4.4.0",
    "history": "^5.3.0",
    "install": "^0.13.0",
    "localforage": "^1.10.0",
    "lucide-react": "^0.428.0",
    "marked": "^14.0.0",
    "match-sorter": "^6.3.4",
    "npm": "^10.8.2",
    "object-path": "^0.11.5",
    "react": "^18.3.1",
    "react-dom": "^18.3.1",
    "react-hook-form": "^7.53.0",
    "react-router-dom": "^6.26.1",
    "sort-by": "^0.0.2",
    "tailwind-merge": "^2.5.2",
    "tailwindcss-animate": "^1.0.7",
    "zod": "^3.23.8",
    "zustand": "^4.5.5"
  },
  "devDependencies": {
    "@eslint/js": "^9.9.0",
    "@types/node": "^22.4.1",
    "@types/react": "^18.3.3",
    "@types/react-dom": "^18.3.0",
    "@vitejs/plugin-react": "^4.3.1",
    "autoprefixer": "^10.4.20",
    "eslint": "^9.9.0",
    "eslint-plugin-react-hooks": "^5.1.0-rc.0",
    "eslint-plugin-react-refresh": "^0.4.9",
    "globals": "^15.9.0",
    "postcss": "^8.4.41",
    "sass-embedded": "^1.78.0",
    "tailwindcss": "^3.4.10",
    "typescript": "^5.5.3",
    "typescript-eslint": "^8.0.1",
    "vite": "^5.4.1"
    
  },`

those are my dependencies.

**Comment by GaetanRdn** [comment]
use `provideExperimentalZonelessChangeDetection()` fixed the error for me ✌🏼

**Comment by shawnweeks** [comment]
Is there a workaround for this yet? Doesn't seem appropriate to revert to a library over a year old at this point.

Updated...

I did find something interesting, if I don't use `127.0.0.1` or `localhost` the error goes away. For example if I use `http://127-0-0-1.sslip.io:4200` to access my app I don't get the error.

**Comment by AliaksandrShustouski** [comment]
This helps me:
"overrides": {
    "@supabase/supabase-js": {
      "@supabase/auth-js": "2.61.0"
    }
  }

**Comment by AugustoMartinez** [comment]
Hi, I'm having the same problem, and I don't know what it could be. In my previous applications, the "overrides" option: {
"@supabase/supabase-js": {
    "@supabase/auth-js": "2.61.0"
  }
}
This solution worked for me, but when I use it in my current application, I now get another error:
`Property 'userStorage' does not exist on type 'SupabaseAuthClientOptions'. `
Do you know what the cause could be? I'm using the latest version of Supabase and Angular.

**Comment by Eliot6001** [comment]
> Hi, I'm having the same problem, and I don't know what it could be. In my previous applications, the "overrides" option: {
> "@supabase/supabase-js": {
>     "@supabase/auth-js": "2.61.0"
>   }
> }
> This solution worked for me, but when I use it in my current application, I now get another error:
> `Property 'userStorage' does not exist on type 'SupabaseAuthClientOptions'. `
> Do you know what the cause could be? I'm using the latest version of Supabase and Angular.


Try running it in a chromium based, updated browser. It worked for me!

**Comment by LIvvieLynn** [comment]
Still had issues w/ supabase 2.61.0 and angular 20.3.0. Was able to resolve by removing zone.js w/ provideZonelessChangeDetection. 

The Promise wrapper fix in the Auth-JS - locks.js fix did not resolve problem for me:
https://github.com/supabase/auth-js/blob/aadf02e63179746a06451f4247a370dfd05740ea/src/lib/locks.ts#L81

I did attempt to debug and noticed back-to-back calls to navigatorLock fn in which acquireTimeout changes from -1 to 0 before first call to navigatorLock was resolved. Does appear that zone.js promises are not being handled properly. 

Link to promise.ts if someone wants to debug further but would recommend upgrading to zoneless if feasible:
https://github.com/angular/zone.js/blob/b11bd466c20dc338b6d4d5bc3e59868ff263778d/lib/common/promise.ts#L122

**Comment by nicetomytyuk** [comment]
> Hi, I'm having the same problem, and I don't know what it could be. In my previous applications, the "overrides" option: { "@supabase/supabase-js": { "@supabase/auth-js": "2.61.0" } } This solution worked for me, but when I use it in my current application, I now get another error: `Property 'userStorage' does not exist on type 'SupabaseAuthClientOptions'. ` Do you know what the cause could be? I'm using the latest version of Supabase and Angular.

The error is caused probably because of inconsistency between supabase-js version and auth-js version, if you remove the override, it should disappar.

Else you can set a static version to supabase-js to 2.49.8 that will work with the overrided auth-js that prevents LockManager issue.

**Comment by craig-at-metashield-io** [comment]
this issue is still present with `"@supabase/supabase-js": "^2.75.0",`  and (still zone-ful) angular 19. 


downgrading `"@supabase/auth-js": "2.61.0"` is not an option if you move to [asymmetric JWT tokens](https://supabase.com/blog/jwt-signing-keys).

**Comment by ekkolon** [comment]
I also encountered this issue on 
- Angular **v21.0.1**
- Supabase **v2.86.0**

I was able to workaround it by:
 
1. Using zoneless Angular - `provideZonelessChangeDetection()` 

**AND**

2. Removing `zone.js` from polyfills in `angular.json` or `project.json`

**Comment by bugprone** [comment]
### Workaround: Pin `@supabase/auth-js` to 2.61.0

We hit this on iOS Safari. `navigatorLock` (introduced in 2.62.0) fails to acquire lock when an existing session is present in localStorage - causing `getSession()` and `INITIAL_SESSION` to hang indefinitely.

Private tabs work fine since there's no persisted session to trigger the lock path.

**Fix:**

```json
{
  "overrides": {
    "@supabase/supabase-js": {
      "@supabase/auth-js": "2.61.0"
    }
  }
}
```

This rolls back to the pre-`navigatorLock` version. Cross-tab session sync still works via `visibilitychange`/`focus` events if you implement it manually.

**Env:** Next.js 16, @supabase/supabase-js 2.87.1, @supabase/ssr 0.8.0

**Comment by JTBrinkmann** [comment]
I could replicate this issue in two circumstances.
I tested it using Angular v19.2.15 using zone.js and supabase-js v2.90.0 on **Firefox** v146 and **Chromium** v143
Going zoneless wasn't possible in my project just yet.

**possible workarounds for users** (besides waiting for a fix): either
1. go zoneless
2. wrap the default locking mechanism to avoid 0ms timeouts:
  ```typescript
  import { createClient, navigatorLock, LockFunc } from '@supabase/supabase-js'
  const lock: LockFunc = (name, acquireTimeout, fn) => {
    return navigatorLock(name, acquireTimeout || 1, fn)
  }
  createClient(url, apiKey, { auth: { lock }})
  ```
3. configure auth-js to a skip any locking all together:
  ```typescript
  createClient(url, apiKey, { auth: { lock: (name, timeout, fn) => fn() }})
  ```
4. configure auth-js to NOT auto refresh tokens:
  ```typescript
  createClient(url, apiKey, { auth: { autoRefreshToken: false } })
  ```


## Analysis
@mandarini `GoTrueClient` sometimes requests multiple locks simultaneously. This usually isn't an issue, as by default the callback **awaits** an available lock, but [`_autoRefreshTokenTick`](https://github.com/supabase/supabase-js/blob/bfa55bb/packages/core/auth-js/src/GoTrueClient.ts#L2959) specifically has a timeout of 0ms, in which case [the `navigatorLock` implementation explicitly sets](https://github.com/supabase/supabase-js/blob/bfa55bb/packages/core/auth-js/src/lib/locks.ts#L128) the [`ifAvailable` flag](https://developer.mozilla.org/en-US/docs/Web/API/LockManager/request#ifavailable), causing the callback to be called with `null` (no lock) instead of awaiting an exclusive lock. This consistently fails. In this specific case (**0ms** timeout) `navigatorLock` even [**throws an (unhandled) error**](https://github.com/supabase/supabase-js/blob/bfa55bb/packages/core/auth-js/src/lib/locks.ts#L156).

I've seen this error being thrown in two circumstances:
1. with default configuration, simply switching to a different tab and back. The [`_onVisibilityChanged` handler](https://github.com/supabase/supabase-js/blob/bfa55bb/packages/core/auth-js/src/GoTrueClient.ts#L3044) acquires a lock **for itself** but also calls `_startAutoRefresh()` which calls `_autoRefreshTokenTick` **right after** (i.e. before the first lock is released). I don't understand how this wouldn't be an issue without zone.js, but didn't investigate further.
2. when calling [`client.auth.onAuthStateChange`](https://github.com/supabase/supabase-js/blob/bfa55bb/packages/core/auth-js/src/GoTrueClient.ts#L2206) in my application code before the auth client's async initialization finished. `onAuthStateChange` acquires a lock to emit the initial session (umm... why?) which won't be released before the *initial* `_autoRefreshTokenTick` call, causing that to fail.

**Possible fixes in auth-js**:
- just log the error instead of throwing it, and
- use a different lock name for `_autoRefreshTokenTick`'s lock
alternatively:
- rework `_onVisibilityChanged` to not use multiple locks on the same name
- make `_autoRefreshTokenTick` use >0ms timeout (as a user I wouldn't want the refresh to randomly/consistently fail because something **unrelated** holding a lock)

@mandarini if needed, I can create a repo with a minimal reproducable example

side note: the type declaration for `SupabaseClientOptions` has an [outdated note](https://github.com/supabase/supabase-js/blob/bfa55bb/packages/core/auth-js/src/lib/types.ts#L122) that no locking is done by default

**Comment by mandarini** [comment]
@JTBrinkmann a reproduction repository would definitely help. Thank you!

**Comment by Polyterative** [comment]
I've been fighting with this for two years, still no solution.

**Comment by mandarini** [comment]
We are working on this @Polyterative . Will come up with solution soon.

**Comment by mandarini** [comment]
Hi everyone, a fix is now available in the canary release:

```bash
npm install @supabase/supabase-js@canary
```

**What changed (PR #2106):** when a lock gets orphaned and isn't released, auth-js now waits up to 5 seconds and then automatically recovers by stealing the stale lock instead of hanging indefinitely.

**This fix has different effects depending on your situation:**

- **If you were experiencing actual hangs** — `getSession()` / auth operations that never complete. This fix should help. Please test the canary and report back.

- **If you're only seeing the "Acquiring an exclusive Navigator LockManager lock immediately failed" console message but auth still works** — this fix does *not* change that. That message comes from the auto-refresh background tick, which intentionally uses a non-blocking lock attempt and skips gracefully when the lock is unavailable. The root cause is zone.js patching the native Promise API used by the Web Locks API. The cleanest fix for Angular apps remains switching to `provideZonelessChangeDetection()` (Angular 18+).

If you were experiencing actual hangs, please test the canary and share your environment (Angular version, zone.js vs zoneless, browser). Thank you for the detailed investigation and your patience!

Also, if anyone has a reproduction repository, that would be SO helpful!

Thank you everyone for contributing to Supabase! You're what makes our tool better! 💚

**Comment by mandarini** [comment]
* retry with `canary` to be specific

**Comment by Polyterative** [comment]
> I've been fighting with this for two years, still no solution.

I managed to fix this by ensuring that I have a single instance across my tabs.Would have never guessed it without Claude.I redid a good amount of my logic to take care and synchronizing different tabs.I no longer see the error.

**Comment by mandarini** [comment]
@Polyterative would you like to try out `2.103.1-beta.0`? It's a release for this PR https://github.com/supabase/supabase-js/pull/2235.

**Comment by cyborgx99** [comment]
> [@Polyterative](https://github.com/Polyterative) would you like to try out `2.103.1-beta.0`? It's a release for this PR [#2235](https://github.com/supabase/supabase-js/pull/2235).



> * retry with `canary` to be specific

It worked for me. The error was gone with this version.

**Comment by mandarini** [comment]
Nice, good to know. In any case, we will not be moving forward with that :( We are working on a different solution. I'll keep you all posted.

**Comment by mandarini** [comment]
Here is the refactor: https://github.com/supabase/supabase-js/pull/2392

Published under `2.107.0-beta.1`.

[Instructions on how to migrate are here](https://github.com/Bewinxed/supabase-js/blob/1a98a572d32bfdaef9b11f95db0dfb028443e622/packages/core/auth-js/migrations/lockless-coordination.md).

Non breaking change, backwards compatible. Please let me know if you try it out, we're trying to evaluate this before publishing.

@supersu-man

**Comment by mandarini** [comment]
[This](https://github.com/supabase/supabase-js/pull/2392) has been merged, going to be released tomorrow in our next minor. Closing
